fem/bilinearform_ext.cpp

```diff
@@ -42,24 +42,76 @@ PABilinearFormExtension::PABilinearFormExtension(BilinearForm *form)
      trialFes(a->FESpace()),
      testFes(a->FESpace())
 {
-   elem_restrict_lex = trialFes->GetElementRestriction(
-                          ElementDofOrdering::LEXICOGRAPHIC);
-   if (elem_restrict_lex)
+   elem_restrict_lex = NULL;
+   int_face_restrict_lex = NULL;
+   bound_face_restrict_lex = NULL;
+}
+
+void PABilinearFormExtension::SetupRestrictionOperators()
+{
+   if (elem_restrict_lex == NULL)
+   {
+      elem_restrict_lex = trialFes->GetElementRestriction(
+                             ElementDofOrdering::LEXICOGRAPHIC);
+      if (elem_restrict_lex)
+      {
+         localX.SetSize(elem_restrict_lex->Height(), Device::GetMemoryType());
+         localY.SetSize(elem_restrict_lex->Height(), Device::GetMemoryType());
+         localY.UseDevice(true); // ensure 'localY = 0.0' is done on device
+      }
+   }
+
+   // Construct face restriction operators only if the bilinear form has
+   // interior or boundary face integrators
+   if (int_face_restrict_lex == NULL && a->GetFBFI()->Size() > 0)
+   {
+      int_face_restrict_lex = trialFes->GetFaceRestriction(
+                                 ElementDofOrdering::LEXICOGRAPHIC, FaceType::Interior);
+      if (int_face_restrict_lex)
```

Comment:
This check seems redundant -- unlike the element restriction case, the face restrictions can never be the identity operator, right?